### PR TITLE
Persist torch.compile artifacts across processes with Mega-cache

### PR DESCRIPTION
**Pull request by danielhanchen**
## Problem

`unsloth_compiled_cache/` caches only the generated source of the patched modules. The expensive part of the first training step -- Dynamo tracing, AOTAutograd partitioning, Inductor codegen and Triton autotuning for the forward and backward graphs -- is redone by every new process unless torch's own on-disk caches happen to be warm, and those default to `/tmp/torchinductor_<user>`, which rarely survives reboots or container restarts. In practice every notebook restart and every fresh container pays the full compile cost again.

## Change

New `unsloth_zoo/compile_cache.py` persists the compile products through the portable Mega-cache API (`torch.compiler.save_cache_artifacts` / `load_cache_artifacts`, torch >= 2.7):

- `unsloth_compile_transformers` calls `megacache_load` before any compiled region runs. The bundle key is an exact fingerprint over torch, CUDA, Triton, transformers, unsloth_zoo, GPU name and architecture, plus the per-model compile kwargs and the Inductor options baked into the generated source. Any mismatch is a silent miss followed by a normal local compile; loading is best-effort and never raises.
- A bundle is written at process exit when compilation happened and no bundle was present. Bundles live under `UNSLOTH_MEGA_CACHE_DIR` (default `~/.cache/unsloth/mega_cache`), one directory per key, with a checksummed manifest so a corrupted bundle is an explicit miss instead of a crash.
- Knobs: `UNSLOTH_MEGA_CACHE=auto|1|0` (auto = load + save-on-miss; 1 also refreshes an existing bundle; 0 = kill switch), `UNSLOTH_MEGA_CACHE_DIR`.
- Also sets `TORCHINDUCTOR_AUTOGRAD_CACHE=1` next to the old `ENABLE_AOT_AUTOGRAD_CACHE` name, which torch >= 2.12 no longer reads, so the backward-graph cache works again on current torch.

Steady-state step time is unchanged; this only affects time to the first optimizer step (TTFS).

## Measurements

B200, torch 2.12.1+cu130, 10-step QLoRA runs from the published notebooks, cold `TORCHINDUCTOR_CACHE_DIR` each time, TTFS measured from `trainer.train()`:

| Model | Cold | With bundle | Speedup | Bundle size |
| --- | --- | --- | --- | --- |
| gpt-oss-20b | 101.0 s | 18.2 s | 5.5x | 54 MB |
| gemma-4-E4B-it | 17.5 s | 9.7 s | 1.8x | 10 MB |

Re-verified before opening this PR with a fresh cold-vs-warm smoke (fresh processes, wiped inductor/triton caches): gpt-oss-20b 3-step QLoRA, cold 103.7 s vs 45.6 s with the bundle loaded (2.3x on the whole 3-step train window, bundle 58 MB, explicit cache hit reported).

## Notes

- Bundles are strictly per-environment: any change in torch/Triton/CUDA build, GPU architecture, transformers version, or the generated-source configuration changes the key and falls back to a normal compile.
- Nothing is shared or uploaded; the cache directory is local.
- `emulate_precision_casts` and other inference-side compile tweaks were deliberately NOT ported to training: they change gradient numerics.

### CODE REVIEW

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/compile_cache.py [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+
+def megacache_save():
+    """Persist this process' compile artifacts. Runs at exit (atexit) or on
+    demand. Saves only when something new would be stored: a fresh bundle on
+    a miss, or a refresh when UNSLOTH_MEGA_CACHE=1. Never raises.
+    """
+    if not (_STATE["armed"] and megacache_is_enabled() and _mega_cache_supported()):
+        return False
+    if _STATE["hit"] and not _refresh_enabled():
+        return False
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/compile_cache.py]: Fixed in 0ec9d9ce: megacache_load resets the hit flag when it computes the new cumulative key, so a hit for an earlier bundle no longer suppresses the exit save after a later miss. Verified with a two-model simulation: A hit, B miss, save proceeds in auto mode.

File: unsloth_zoo/compile_cache.py [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+        if not result or not result[0]:
+            _log("nothing to save (no compile artifacts recorded)")
+            return False
+        data = result[0]
+        directory, bundle_path, manifest_path = _bundle_paths(key)
+        os.makedirs(directory, exist_ok = True)
+        temporary_path = bundle_path + ".tmp"
+        with open(temporary_path, "wb") as file:
+            file.write(data)
+        os.replace(temporary_path, bundle_path)
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/compile_cache.py]: Fixed in 0ec9d9ce: saves are now main-rank only (RANK, falling back to LOCAL_RANK), and both bundle and manifest write through pid-unique temp files with os.replace, so no writer can interleave or truncate another's bytes and the manifest checksum always matches the bundle it was computed from.

File: unsloth_zoo/compiler.py [hunk 3]
```diff
@@ -3448,6 +3448,44 @@ def replaced_tqdm(*args, **kwargs):
...
+                "compile_function_calls": compile_function_calls,
+                "fuse_lm_head"          : fuse_lm_head,
+                "gradient_checkpointing": gradient_checkpointing,
+                "manual_replacements"   : manual_replacements,
+                "fast_lora_forwards"    : fast_lora_forwards,
+                "fast_residual_stream"  : fast_residual_stream,
+                "accurate_accumulation" : accurate_accumulation,
+                "fullgraph"             : fullgraph,
+                "disable"               : disable,
+                "return_logits"         : return_logits,
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/compiler.py]: Fixed in 0ec9d9ce: the key now uses the effective fullgraph and return_logits values with the same env-wins precedence the code applies below, and includes UNSLOTH_RETURN_HIDDEN_STATES since the generated forwards branch on it. Different effective modes now produce different keys.

File: unsloth_zoo/compile_cache.py [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+
+def _refresh_enabled():
+    return os.environ.get("UNSLOTH_MEGA_CACHE", "auto").strip().lower() in ("1", "on", "true", "yes")
+pass
+
+
+def _cache_root():
+    root = os.environ.get("UNSLOTH_MEGA_CACHE_DIR", "")
+    if root == "":
+        root = os.path.join(os.path.expanduser("~"), ".cache", "unsloth", "mega_cache")
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/compile_cache.py]: Fixed in 0ec9d9ce: the env value is expanduser'd before use.

File: unsloth_zoo/compile_cache.py [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+    return root
+pass
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/compile_cache.py:107]: Addressed in 0ec9d9ce for the real failure modes: the env override is expanduser'd, and an unresolvable home (no HOME in a restricted container) now disables the cache cleanly instead of creating a literal tilde directory in the CWD; all load/save paths already never raise. I did not adopt the shared-tmp fallback: writing a world-visible cache under /tmp is exactly the symlink/ownership surface the suggestion then has to defend against, and disabling is safer than degrading.

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/compile_cache.py [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+
+    ``compile_kwargs`` are the unsloth_compile_transformers arguments that
+    change the GENERATED SOURCE (and therefore the traced graphs);
+    ``torch_compile_options`` is the Inductor options dict baked into the
+    generated ``@torch.compile`` decorators.
+    """
+    payload = {
+        "model_type": str(model_type),
+        "compile_kwargs": {k: compile_kwargs[k] for k in sorted(compile_kwargs)},
+        "torch_compile_options": {k: torch_compile_options[k] for k in sorted(torch_compile_options)},
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/compile_cache.py:177]: Fixed in 0c1badb3 via content-based saving rather than config key dimensions: torch records artifacts on cache hits too, so save_cache_artifacts serializes loaded plus newly compiled entries. The save now compares that content against the bundle on disk and writes whenever it differs, so the 8B-bundle-plus-70B-run case persists a merged bundle covering both checkpoints and the next 70B run hits. Config dims in the key would fragment bundles per checkpoint without fixing anything torch's internal cache keys do not already disambiguate. Simulated: hit plus new artifacts saves the superset; pure hit skips.

File: unsloth_zoo/compile_cache.py [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+        "gpu_capability": None,
+    }
+    try:
+        import torch
+        fingerprint["torch"] = str(torch.__version__)
+        fingerprint["cuda"] = str(torch.version.cuda)
+        if torch.cuda.is_available():
+            fingerprint["gpu_name"] = torch.cuda.get_device_name(0)
+            capability = torch.cuda.get_device_capability(0)
+            fingerprint["gpu_capability"] = f"sm_{capability[0]}{capability[1]}"
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/compile_cache.py:144]: Fixed in 0c1badb3: the fingerprint reads torch.cuda.current_device(). Verified on a live two-GPU box with set_device(1): the fingerprint reports device 1's name and capability.

File: unsloth_zoo/compiler.py [hunk 8]
```diff
@@ -3448,6 +3448,55 @@ def replaced_tqdm(*args, **kwargs):
...
+        # Env vars override these arguments below (and the generated forwards
+        # branch on UNSLOTH_RETURN_HIDDEN_STATES), so key on the EFFECTIVE
+        # values or one mode's bundle would be a false hit for another.
+        _effective_fullgraph = os.environ.get(
+            "UNSLOTH_FULLGRAPH", "1" if fullgraph else "0"
+        ) == "1"
+        _effective_return_logits = os.environ.get(
+            "UNSLOTH_RETURN_LOGITS", "1" if return_logits else "0"
+        ) == "1"
+        _return_hidden_states = os.environ.get("UNSLOTH_RETURN_HIDDEN_STATES", "0") == "1"
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/compiler.py:3469]: Fixed in 0c1badb3 by the same content-based save: a mode flipped after from_pretrained compiles new graphs, the exit save detects the content difference and persists them into the bundle under the load-time key, and the next run loads that superset and hits regardless of when the env var is set. Keying save-time env values instead would strand the artifacts under a key the next run's load never computes.

File: unsloth_zoo/compile_cache.py [hunk 9]
```diff
@@ -0,0 +1,328 @@
...
+    """
+    if not (_STATE["armed"] and megacache_is_enabled() and _mega_cache_supported()):
+        return False
+    if _STATE["hit"] and not _refresh_enabled():
+        return False
+    # Distributed launches: every rank records the same artifacts; concurrent
+    # writers could leave the manifest checksum pointing at another rank's
+    # bytes, so only the main rank persists the bundle.
+    if os.environ.get("RANK", os.environ.get("LOCAL_RANK", "0")) != "0":
+        return False
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/compile_cache.py]: Fixed in 0c1badb3: the writer gate is now per node (LOCAL_RANK first, RANK fallback), so every node's main rank persists its node-local cache. Shared roots remain safe: writes go through pid-unique temp files with atomic replace, and the manifest is written the same way, so concurrent node writers cannot interleave or mismatch checksums.

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/compile_cache.py [hunk 10]
```diff
@@ -0,0 +1,342 @@
...
+        data = result[0]
+        directory, bundle_path, manifest_path = _bundle_paths(key)
+        if bundle_path is None:
+            return False
+        new_sha = hashlib.sha256(data).hexdigest()
+        if not _refresh_enabled():
+            try:
+                with open(manifest_path, "r") as file:
+                    if json.load(file).get("sha256") == new_sha:
+                        _log("bundle unchanged; skipping save")
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/compile_cache.py]: Genuine, fixed in 3afb86d0. Saves now deserialize the on-disk bundle and fold its artifacts back into torch's recorder (deduplicated by artifact value equality) before serializing, so the written bundle is always the union of disk and process state. Verified with a two-process test: record A+B, then a partial run recording only C ends with a bundle containing A, B and C.

File: unsloth_zoo/compile_cache.py [hunk 11]
```diff
@@ -0,0 +1,342 @@
...
+            "created": time.time(),
+            "bytes": len(data),
+            "sha256": new_sha,
+            "env": _environment_fingerprint(),
+            "models": _STATE["model_keys"],
+        }
+        temporary_manifest = f"{manifest_path}.tmp.{os.getpid()}"
+        with open(temporary_manifest, "w") as file:
+            json.dump(manifest, file, indent = 2, sort_keys = True, default = str)
+        os.replace(temporary_manifest, manifest_path)
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/compile_cache.py:371]: Fixed in 3afb86d0 without a lock: the bundle file is content-addressed (megacache-<sha16>.bin) and the manifest names it. Different content yields different file names, so a manifest can never checksum-match another writer's bytes; the manifest replace stays atomic and whichever lands last points at its own complete file. Superseded bundles are unlinked best-effort after the switch.

File: unsloth_zoo/compiler.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/compiler.py:3469]: The load-time key is intentional and safe for post-load mode flips. Both modes of a checkpoint resolve to the same bundle key (the key is computed at load, before the training loop flips the env), so GRPO's hidden-state compiles land in the same bundle as the normal-training artifacts, and with union saves (3afb86d0) that accumulation is lossless. On load, torch's internal per-artifact cache keys disambiguate the modes, so a mode the bundle has not seen is a silent first-compile that then merges in, never a wrong artifact. Keying at save time would split one checkpoint's artifacts across bundles keyed by whatever mode happened to be live at exit, which fragments the cache without adding correctness. The effective-env keying added earlier covers users who set these vars before load, where they change what the generated source itself looks like.

**Comment by danielhanchen** [comment]
@codex review
